node/pkg/governor/governor.go

```diff
@@ -656,34 +796,83 @@ func computeValue(amount *big.Int, token *tokenEntry) (uint64, error) {
 	return value, nil
 }
 
-func (gov *ChainGovernor) TrimAndSumValueForChain(ce *chainEntry, startTime time.Time) (sum uint64, err error) {
-	sum, ce.transfers, err = gov.TrimAndSumValue(ce.transfers, startTime)
-	return sum, err
+// TrimAndSumValueForChain calculates the `sum` of `Transfer`s for a given chain `emitter`. In effect, it represents a
+// chain's "Governor Usage" for a given 24 hour period.
+
+// This sum may be reduced by the sum of 'flow cancelling' transfers: that is, transfers of an allow-listed token
+// that have the `emitter` as their destination chain.
+// The resulting `sum` return value therefore represents the net flow across a chain when taking flow-cancelling tokens
+// into account. Therefore, this value should never be less than 0 and should never exceed the "Governor limit" for the chain.
+//
+// As a side-effect, this function modifies the parameter `emitter`, updating its `transfers` field so that it only includes
+// filtered `Transfer`s (i.e. outgoing `Transfer`s newer than `startTime`).
+//
+// SECURITY Invariant: The `sum` return value should never be less than 0
+// SECURITY Invariant: The `sum` return value should never exceed the "Governor limit" for the chain
+func (gov *ChainGovernor) TrimAndSumValueForChain(emitter *chainEntry, startTime time.Time) (sum uint64, err error) {
+	// Sum the value of all outgoing transfers
+	var sumOutgoing int64
+	sumOutgoing, emitter.transfers, err = gov.TrimAndSumValue(emitter.transfers, startTime)
+	if err != nil {
+		return 0, err
+	}
+
+	// Return early if the sum is not positive as it cannot exceed the daily limit.
+	// In this case, return 0 even if the sum is negative.
+	if sumOutgoing <= 0 {
+		return 0, nil
+	}
+
+	sum = uint64(sumOutgoing)
+	if sum > emitter.dailyLimit {
+		return 0, fmt.Errorf(
+			"invariant violation: calculated sum %d exceeds Governor limit %d",
+			sum,
+			emitter.dailyLimit,
+		)
+	}
+
+	return sum, nil
+
 }
 
-func (gov *ChainGovernor) TrimAndSumValue(transfers []*db.Transfer, startTime time.Time) (uint64, []*db.Transfer, error) {
+// TrimAndSumValue iterates over a slice of transfer structs. It filters out transfers that have Timestamp values that
+// are earlier than the parameter `startTime`. The function then iterates over the remaining transfers, sums their Value,
+// and returns the sum and the filtered transfers.
+// The `transfers` slice must be sorted by Timestamp. We expect this to be the case as transfers are added to the
+// Governor in chronological order as they arrive.
```

Comment:
Is this still true with the flow cancelling transfers since they come from a different chain with it's own timestamp that might drift from the timestamp of the destination chain (being flow cancelled with)?

Reply:
Yep, the `TimeStamp` is generated by this file in the method `ProcessMsg` and `ProcessMsgForTime`; it isn't read from the transfer itself.
I'll modify the code comment here and in the db.Transfer struct to make this more clear.